### PR TITLE
[WIP] Refactor the build_system feature to merge it into the fallback feature

### PR DESCRIPTION
**Pull request by Lattay**
This pull request provide an implementation of the ideas discussed in #2.
For short, it aims at providing a homogeneous interface for simplest tasks regardless of the build system.
The interface is mirrored from the existing interface of the fallback system where default targets were defined depending on the filetype.
In effect this PR use the existing fallback code to implement the build system code.

To make clear that targets may be something else that a build target, they are renamed subcommands.

TODO:
- [x] refactor to use only subcommands and not prefix + target
- [x] allow for user customization
- [x] update inline documentation
- [x] update default build systems definitions
- [x] provide `:Build init`
- [x] update doc/build.txt
- [x] update README.md
- [x] test different build systems

Bonus:
- [x] provide `:Build info`
- [x] add internal `:Build help`
- [x] provide a `:Build test`

### CODE REVIEW

**Comment by Lattay** [comment]
@AlxHnr, I would need some help to make sure the updated build systems are fine because I never used most of them, or not as a developer.
Dune and Make seems ok.
I ran CMake on the Raylib project and it compiled well, but I don't know how sensible is the `run` command.
I think the Autotools config is OK but I only worked on a single project using it, so I don't have too much info on common file architectures used with it.
I tried Cargo on a hello_world type project, it works.
I know nothing about Maven and DUB, and couldn't even write a line of D or Java.

**Comment by Lattay** [comment]
@AlxHnr I think this is good to go, if you want to review the doc and code.

**Comment by AlxHnr** [comment]
I hope to find some time this weekend to look at your work.

> I know nothing about Maven and DUB, and couldn't even write a line of D or Java.

Same for me. I only tried dub on a simple hello-world project five years ago. I don't know if these commands still work today. Only thing I know about Maven is that the build command should use the `compile` target.

`:BuildRefresh` should be moved into its own issue. Multi-build system support has its caveats and should be solved separately. For now, I prefer `:Build` to be as stateless as possible. What if I switch from my current CMake project to a Maven project, but `s:current_build_system` still references CMake? A project like Neovim contains both a CMake file and a normal Makefile. How do we detect which build system is the "main" one?

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt]: I feel that tagging `%PATH%` is a little excessive and _may_ conflict with other plugins. Its such a short tag.

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]: You don't have to wrap at 80 characters. 100 is fine.

File: doc/build.txt
Comment (Lattay) [reply doc/build.txt]: You are right. I think it is important to be able to reach it through help, what do you think about build_%PATH% as a tag ?
Or I simply remove the tags.

File: autoload/build.vim
Comment (Lattay) [reply autoload/build.vim]: Ok, I'll fix that

**Comment by Lattay** [comment]
> Same for me. I only tried dub on a simple hello-world project five years ago. I don't know if these commands still work today. Only thing I know about Maven is that the build command should use the `compile` target.

Well, at least I should change this to use compile.

> `:BuildRefresh` should be moved into its own issue. Multi-build system support has its caveats and should be solved separately. For now, I prefer `:Build` to be as stateless as possible. 

Yes, I anticipated a bit too much. I introduced a caching Build thinking about your suggestion of `Build chdir`, and so I had to introduce `BuildRefresh` to force update but then I realized that `Build chdir` and other linked features should be a separated PR but I did not reverse the changes introducing `BuildRefresh`.

> What if I switch from my current CMake project to a Maven project, but `s:current_build_system` still references CMake?

It is the reason why I added `BuildRefresh`  it would detect the change when called.

>  A project like Neovim contains both a CMake file and a normal Makefile. How do we detect which build system is the "main" one?

The problem already exists independently of the changes I introduced. Your algo will choose the first encountered.
I thought about adding a `:Build set_build_system` command to manually force a choice, but it should not be in this PR either.

**Comment by Lattay** [comment]
To prevent stupid conflicts in the dune file I propose that it is kept in alphabetical order of the executables names, it should prevent most problems.

**Comment by Lattay** [comment]
Should be ok now if you accept the tags `build_%MAGIC%`.

File: doc/build.txt
Comment (AlxHnr) [reply doc/build.txt]: I would prefer to simply remove the tags. In order to search for `:help build_%PATH%` you would already have to know what `%PATH%` is.

**Comment by AlxHnr** [comment]
This plugin was not always stateless. It had a `build#setup()` function, which was called each time a file was created or loaded. After every change to the configuration it was necessary to go trough all open buffers and run `:e` to force re-detection. Thus the feature was removed. I also removed a [chdir feature](https://github.com/AlxHnr/build.vim/commit/8165a69778389a66fdadc22ac462095d6c7d65a3) feature based on this state.

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt]: You should add a placeholder to this example. This shows that commands can do way more and makes people keep reading.

```suggestion
  \       'run'  : 'foomake cmd ./%HEAD%',
```

File: README.md
Comment (AlxHnr) [inline README.md]: Not relevant to this PR.

File: README.md
Comment (AlxHnr) [inline README.md]: The commands `BuildInit` and `BuildInfo` only exist for backwards compatibility. Leaving them in the README implies they should still be used.

File: README.md
Comment (AlxHnr) [inline README.md]: Most of these comments don't say anything.

File: README.md
Comment (AlxHnr) [inline README.md]: Typo: `imply` -> `implies`

You should move this commend and the python3 block below `:Build run`.

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]:
```suggestion
  \        'clean' : 'cmake --build ./build/ -- clean',
```

This is what I expect from the command in order to rebuild the project. Otherwise it would _de-initialize_ the project, which is to aggressive.

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]:
```suggestion
  \        'build' : 'mvn compile',
```

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim:144]: Does the run command care about the output of the build command in this case?

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]: Is this a typo? If its intentional, you should pick a better example.

```suggestion
"   s:relative_to('/a/b/c/d/e/f/g', '../d/e')
```

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]: You don't need `l:initial_wd`.

```suggestion
  lchdir -
```

Please use full command names for consistency. E.g. `exec` -> `execute`, `lcd` -> `lchdir`.

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]:
```suggestion
" If the build_system is a fallback, both g:build#languages and s:language_cmds
```

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim:245]: Why does this function take a dictionary instead of two arguments?

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]: `extra_args` are not escaped.

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]: Vim globbing contains to much magic. Matching full, explicit file names is simpler and gets the job done.

```suggestion
      if filereadable(a:dir . '/' . l:build_file)
```

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]: This refers to the removed `:BuildRefresh` commands.

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]: This variable should've been removed together with `:BuildRefresh`.

```suggestion
    return {'name': &filetype, 'fallback': v:true, 'path': '.'}
```

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim]: This variable should've been removed together with `:BuildRefresh`.

Comment (AlxHnr) [reply autoload/build.vim]: I see it is still used in many places in this file. Remove those too.

File: autoload/build.vim
Comment (AlxHnr) [inline autoload/build.vim:478]: `:Build info` and `:Build help` should be merged into one function and always do the same thing. This is better for discoverability. Imo `:Build help` should be even removed completely. We have `:help build` for that.

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt]: Remove `:BuildInit` and `:BuildInfo` completely from this help page. The commands are deprecated now and only exist for backwards compatibility.

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt]: Don't forget to change this CMake command example.

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt]: Globs should be removed here too.

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt:153]: Add a tag to this section so they can be referenced in the documentation of `g:build#languages.`.

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt]: Reference the main placeholder tag here.

**Review by AlxHnr — CHANGES_REQUESTED** [review]
Im not sure for the defaults of most run targets. Most files don't end up directly in the `build/` directory, so `run: './build/%HEAD%` doesn't make sense. What if a file named `test/foo/bar/mytest.cpp` ends up in `build/some_subdir/test/foo/bar/mytest`?

I don't see how this can be solved in a generic way. Imo the defaults should do the same as `:Build do run`, but can be customized by `g:build#systems`.

File: autoload/build.vim
Comment (AlxHnr) [reply autoload/build.vim]: Append them directly to `a:cmd` with a whitespace in between.

File: README.md
Comment (Lattay) [reply README.md]: Indeed, I'll fix that

File: README.md
Comment (Lattay) [reply README.md]: Yes, you are right.

File: autoload/build.vim
Comment (Lattay) [reply autoload/build.vim:144]: No it doesn't. To run the output of the build you would simply use something like `./%HEAD%`.

File: autoload/build.vim
Comment (Lattay) [reply autoload/build.vim]: Not a typo, but I have to admit it is not the best example. I'll try to find something better.

File: autoload/build.vim
Comment (Lattay) [reply autoload/build.vim:245]: It is a `build_system` dictionary as returned by `build#get_current_build_system`. I can change that but for the caller it seems easier to use the full dict and it may be better if `gather_commands` needs more info on the build system.

File: autoload/build.vim
Comment (Lattay) [reply autoload/build.vim]: Again it was too much anticipation.
There is the case of Chicken Scheme *chicken-install* which is effectively a build system and use a `name_of_project.egg` file.
Since the change to support glob was so simple it seemed ok to me.
What kind of magic is a problem to you ?

File: autoload/build.vim
Comment (Lattay) [reply autoload/build.vim]: Indeed, sorry about that.

File: autoload/build.vim
Comment (Lattay) [reply autoload/build.vim:478]: I understand for the command but I think it is still nicer to have separated functions internally because it enables a better help message when using a command that is not available.

File: doc/build.txt
Comment (Lattay) [reply doc/build.txt]: What would you expect ? I am not a big user of CMake, but I think most of the time there is no clean target and you don't need to clean unless you want to trash everything and restart from scratch.

**Comment by AlxHnr** [comment]
It seems like you pushed the changes to your master branch, while this merge request still tracks `Lattay:refactor`. You have to either push these commits to this branch or switch this merge requests branch. I've unresolved all discussions and will resolve them myself when you push your changes.

**Comment by Lattay** [comment]
Oh ! Yes sorry about that.

File: README.md
Comment (AlxHnr) [inline README.md]:
```suggestion
:Build clean
```

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt]: You can remove this entire paragraph. `BuildInit` and `BuildInfo` are deprecated.

File: doc/build.txt
Comment (AlxHnr) [inline doc/build.txt]:
```suggestion
                                                             *build_placeholders*
```

File: doc/build.txt
Comment (AlxHnr) [reply doc/build.txt]: `cmake --build ./build/ -- clean`

Comment (AlxHnr) [reply doc/build.txt]: This is consistent with `s:build_systems`. Almost every project has a `clean` target.

**Review by AlxHnr — CHANGES_REQUESTED** [review]
The last typos.

File: autoload/build.vim
Comment (AlxHnr) [reply autoload/build.vim]: I see the need to support `*.egg`. But I still have mixed feelings. Glob allows recursive searches via `glob('**')` and even run shell commands:

```vim
glob('`find -iname "*.cpp"`')
```

By default its behaviour is dependent on `wildignore` and `suffixes`.

Comment (AlxHnr) [reply autoload/build.vim]: What if `a:dir` contains special characters? E.g. the PWD is `/home/user/**`, which is a legal directory name, but gets interpreted by `glob()`. It could be worked around by lchdir + `glob('.' ...)` but thats hacky.

**Comment by AlxHnr** [comment]
The only thing left is the `glob()` discussion, which GitHub seems to hide. I will fix remaining typos myself after merging.

**Comment by AlxHnr** [comment]
Are you sure you aren't better off maintaining your own fork? I believe you've got a lot of great ideas and energy, but I can't promise you to have time in the future to collaborate. You will very likely end up with your own fork anyways if you want more features. It took me a week to respond to your pull request and I don't want to waste your time any more. We started with adding a few _dune_ commands and ended up with half a rewrite. I'm not sure I want to maintain all of this.

I don't expect you to do free work. I don't expect anything at all and just don't care. This plugin already does everything I need and upstreaming your ideas won't have any impact since no one uses _build.vim_.

Its free and MIT licensed. Do whatever you want with it.

**Comment by Lattay** [comment]
Hey, I have actually a lot of free time right now because I am between two jobs, but I am sorry I took so much of your time, I can understand that you were not prepared for my contributions, which have been very opinionated as I always am.
Even if no one uses this project it was interesting to me to contribute with a different point of view and I got carried away more than once.
I am ok with fixing more things before you merge if you want, I don't want to be invasive either.
I will probably maintain my own fork indeed for other features, but the features I implemented in this PR seemed profitable to anyone and matched my vision of open-source to submit them to you.

Anyway, thanks for everything :+1:

**Comment by AlxHnr** [comment]
I've rebased and merged your changes manually to master. I had to clean up some of the formatting and wording, but it seems to work well. Thanks a lot for your contribution. Some of the issues with `glob()` have been solved by switching to `globpath()`.

**Pull request closed without merging**
